mm2src/coins_activation/src/eth_with_token_activation.rs

```diff
@@ -0,0 +1,235 @@
+            EthActivationV2Error::InvalidSwapContractAddr(e) => {
+                EnablePlatformCoinWithTokensError::InvalidSwapContractAddr(e)
+            },
+            EthActivationV2Error::InvalidFallbackSwapContract(e) => {
+                EnablePlatformCoinWithTokensError::InvalidFallbackSwapContract(e)
+            },
+        }
+    }
+}
+
+impl TryFromCoinProtocol for CoinProtocol {
+    fn try_from_coin_protocol(proto: CoinProtocol) -> Result<Self, MmError<CoinProtocol>>
+    where
+        Self: Sized,
+    {
+        Ok(proto)
+    }
+}
+
+pub struct Erc20Initializer {
+    platform_coin: EthCoin,
+}
+
+impl From<Erc20TokenActivationError> for InitTokensAsMmCoinsError {
+    fn from(error: Erc20TokenActivationError) -> Self {
+        match error {
+            Erc20TokenActivationError::InternalError(e) => InitTokensAsMmCoinsError::Internal(e),
+            Erc20TokenActivationError::CouldNotFetchBalance(e) => InitTokensAsMmCoinsError::CouldNotFetchBalance(e),
+        }
+    }
+}
+
+#[async_trait]
+impl TokenInitializer for Erc20Initializer {
+    type Token = EthCoin;
+    type TokenActivationRequest = Erc20TokenActivationRequest;
+    type TokenProtocol = Erc20Protocol;
+    type InitTokensError = Erc20TokenActivationError;
+
+    fn tokens_requests_from_platform_request(
+        platform_params: &EthWithTokensActivationRequest,
+    ) -> Vec<TokenActivationRequest<Self::TokenActivationRequest>> {
+        platform_params.erc20_tokens_requests.clone()
+    }
+
+    async fn enable_tokens(
+        &self,
+        activation_params: Vec<TokenActivationParams<Erc20TokenActivationRequest, Erc20Protocol>>,
+    ) -> Result<Vec<EthCoin>, MmError<Erc20TokenActivationError>> {
+        let mut tokens = vec![];
+        for param in activation_params {
+            let token: EthCoin = self
+                .platform_coin
+                .initialize_erc20_token(param.activation_request, param.protocol, param.ticker)
+                .await?;
+            tokens.push(token);
+        }
+
+        Ok(tokens)
+    }
+
+    fn platform_coin(&self) -> &EthCoin { &self.platform_coin }
+}
+
+#[derive(Clone, Deserialize)]
+pub struct EthWithTokensActivationRequest {
+    #[serde(flatten)]
+    platform_request: EthActivationV2Request,
+    erc20_tokens_requests: Vec<TokenActivationRequest<Erc20TokenActivationRequest>>,
+}
+
+impl TxHistory for EthWithTokensActivationRequest {
+    fn tx_history(&self) -> bool { self.platform_request.tx_history }
+}
+
+impl TokenOf for EthCoin {
+    type PlatformCoin = EthCoin;
+}
+
+impl RegisterTokenInfo<EthCoin> for EthCoin {
+    fn register_token_info(&self, token: &EthCoin) {
+        let fut = self.add_erc_token_info(token.ticker().to_string(), Erc20TokenInfo {
+            token_address: token.erc20_token_address().unwrap(),
+            decimals: token.decimals(),
+        });
+        futures::executor::block_on(fut);
```

Comment:
Calling `block_on` may block the async runtime thread, preventing other tasks from execution. https://docs.rs/tokio/latest/tokio/task/index.html#blocking-and-yielding. It may become a big problem if called for a potentially long-running IO task. 

If `add_erc_token_info` was really required to be async, then `register_token_info` should have been made `async` too. But I guess `AsyncMutex` can be replaced with std or parking_lot mutex and ideally used with `block_in_place` or `async_blocking`. However, we didn't experience issues with them in async context if they are released in place/not locked for a relatively long time.

Reply:
True, it blocks the caller thread. If user sends very big token activation array on `enable_eth_with_tokens` rpc method, then it might become long running task.

If I do `register_token_info` as async, then other token info structs needs to be `AsyncMutex` as well. But looks like it's worth for it.

Reply:
> What do you think?

Seems a bit tricky 🙂 

Can we maybe use non-async mutex and clone `erc20_tokens_infos` when required? As I can see, info is cloned anyway here:
https://github.com/KomodoPlatform/atomicDEX-API/blob/be591297f39950347f870289067c6b5577da656f/mm2src/coins/eth.rs#L2553 It's actually a data-race introduce, but it won't break the current code since activation process is in fact sequential.

Also, is there a way to refactor it somehow so `AsyncMutex` and `.clone()` will be *both* not required? 🙂

Reply:
> It's actually a data-race introduce, but it won't break the current code since activation process is in fact sequential.

True, my bad. Will get rid of cloning, it easy to not use clone here.


With non-async mutex, since `get_erc_tokens_infos` fn returns `HashMap` behind the `MutexGuard`, cloning it won't be enough. We still can't pass it into `get_token_balance_by_address` because it's awaiting.

In order to not use `AsyncMutex`, I have to get `HashMap<String, Erc20TokenInfo>` without `MutexGuard` from `get_erc_tokens_infos`. But that means `erc20_tokens_infos` type is Mutexed but not using it's protection/benefits 🙂

Reply:
>  this implementation drops the mutex guard and then returns the value which kills the benefit of Mutex on reading the value. So it's only write-safe mutexed. If you think this implementation is valid, I can do that too so AsyncMutex can be replaced with std one.

I don't consider it as fully "valid" since yes, it will be only write-safe mutex. But it might be still better than locking `erc20_tokens_infos` for the execution time of possibly dozens of RPC requests. So, if we implement `my_balance` 2.0 with `with_tokens` argument as suggested here https://github.com/KomodoPlatform/atomicDEX-API/issues/1006#issue-938826711, we might get new `enable_erc20` requests stuck if made concurrently with balance call with all tokens. 

But what's good: we can follow the same approach as `get_spl_tokens_infos` *for now*, and think how we can do it better when (and *if*) `my_balance` 2.0 will be implemented 🙂 Because again: `get_erc_tokens_infos` is used only during activation process for now, which is in fact sequential and won't suffer from race condition introduced by "only write-safe" mutex.

Reply:
PS, but it's a trade-off, of course. We should note it for the future refactorings: https://github.com/KomodoPlatform/atomicDEX-API/issues/1006#issuecomment-1199163387